### PR TITLE
DRILL-3927: use OutOfMemoryException in more places

### PR DESCRIPTION
**Pull request by cwestin**
Some code cleanup required for the upcoming introduction of the rewritten
direct memory allocator. Chiefly the introduction of OutOfMemoryException
handling in a few spots, the use of DrillAutoCloseables for some allocator
close() calls, and some other minor cleanup.
- removed the exclusion of exec/work from the jdbc-all jar, because it now
  depends on OutOfMemoryException, and that is derived from
  FragmentSetupException and Foreman Exception

Unit tests pass
Regression suite passes, except for
- a couple of the known "Exceeded timeout (10000) while waiting send intermediate work fragments to remote nodes."
- a couple of the known "Selected column 'dir0' must have name 'columns' or must be plain '*'"

### CODE REVIEW

File: common/src/main/java/org/apache/drill/common/DeferredException.java [hunk 1]
```diff
@@ -47,6 +75,15 @@ public void addException(final Exception exception) {
...
+        if (exceptionSupplier == null) {
+          this.exception = exception;
+        } else {
+          this.exception = exceptionSupplier.get();
+          if (this.exception == null) {
+            this.exception = new RuntimeException("Missing root exception");
+          }
+          this.exception.addSuppressed(exception);
+        }
         this.exception = exception;
```
Comment (adeneche) [inline common/src/main/java/org/apache/drill/common/DeferredException.java]: should we remove this line ?

File: common/src/main/java/org/apache/drill/common/DeferredException.java [hunk 2]
```diff
@@ -123,6 +160,9 @@ public void suppressingClose(final AutoCloseable autoCloseable) {
         autoCloseable.close();
       } catch(final Exception e) {
         addException(e);
+      } catch(final Error e) {
+        logger.warn("caught Error {}", e);
```
Comment (adeneche) [inline common/src/main/java/org/apache/drill/common/DeferredException.java]: why log the error here ? we have a JIRA to try and do reduce error logging to the minimum [DRILL-2692](https://issues.apache.org/jira/browse/DRILL-2692)

File: common/src/main/java/org/apache/drill/common/DrillAutoCloseables.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+   * {@see java.lang.Exception}. This wraps the close() call with a
+   * try-catch that will rethrow an Exception wrapped with a
+   * {@see java.lang.RuntimeException}, providing a way to call close()
+   * without having to do the try-catch everywhere or propagate the Exception.
+   *
+   * @param closeable the AutoCloseable to close; may be null
+   * @throws RuntimeException if an Exception occurs; the Exception is
+   *   wrapped by the RuntimeException
+   */
+  public static void closeNoChecked(final AutoCloseable autoCloseable) {
```
Comment (adeneche) [inline common/src/main/java/org/apache/drill/common/DrillAutoCloseables.java]: can't we just add this to org.apache.drill.common.AutoCloseables ? it seems to serve the same purpose

File: common/src/main/java/org/apache/drill/common/config/DrillConfig.java [hunk 4]
```diff
@@ -278,9 +265,9 @@ public String toString() {
     return this.root().render();
   }
 
-  public static void main(String[] args)  throws Exception {
+  public static void main(String[] args) {
```
Comment (adeneche) [inline common/src/main/java/org/apache/drill/common/config/DrillConfig.java]: I think it's safe to just remove the main method altogether.

File: common/src/main/java/org/apache/drill/common/DrillAutoCloseables.java (hunk 3 above)
Comment (cwestin) [inline common/src/main/java/org/apache/drill/common/DrillAutoCloseables.java]: My plan was to merge those two files once I'm done -- and only leave one of them. They were created in two different sets of work for the same purpose. Can we do that later?

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java [hunk 5]
```diff
@@ -369,6 +378,10 @@ public void queryCompleted(QueryState state) {
 
     private void fail(Exception ex) {
       logger.debug("Submission failed.", ex);
+      final Throwable cause = ex.getCause();
+      if (cause != null) {
+        logger.debug("Submission failure cause.", ex.getCause());
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: same comment related to [DRILL-2692](https://issues.apache.org/jira/browse/DRILL-2692)

File: common/src/main/java/org/apache/drill/common/DeferredException.java (hunk 1 above)
Comment (cwestin) [inline common/src/main/java/org/apache/drill/common/DeferredException.java]: Yes, that would be a bug. Done.

File: common/src/main/java/org/apache/drill/common/DeferredException.java (hunk 2 above)
Comment (cwestin) [inline common/src/main/java/org/apache/drill/common/DeferredException.java]: OK, removed.

**Comment by jacques-n** [comment]
I'm against having a caught exception for OutOfMemory. This should be a RuntimeException since it can happen anywhere. Why did you choose to use this instead of the Runtime version?

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ImplCreator.java [hunk 6]
```diff
@@ -45,6 +45,7 @@
 public class ImplCreator {
   private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ImplCreator.class);
 
+  private RootExec root = null;
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ImplCreator.java]: `root` is never used, right ?

File: common/src/main/java/org/apache/drill/common/config/DrillConfig.java (hunk 4 above)
Comment (cwestin) [inline common/src/main/java/org/apache/drill/common/config/DrillConfig.java]: Done.

File: exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java (hunk 5 above)
Comment (cwestin) [inline exec/java-exec/src/main/java/org/apache/drill/exec/client/DrillClient.java]: Removed.

**Comment by cwestin** [comment]
I don't understand your comment, Jacques. OutOfMemoryException exists, and is a lot more specific than RuntimeException. Otherwise, what is OutOfMemoryException for? There are no comments in the file, so I'm forced to guess.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ImplCreator.java (hunk 6 above)
Comment (cwestin) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/ImplCreator.java]: Apparently not. Removed.

File: exec/jdbc-all/pom.xml [hunk 7]
```diff
@@ -420,7 +420,6 @@
                <exclude>org/apache/drill/exec/server/rest/**</exclude>
                <exclude>org/apache/drill/exec/rpc/data/**</exclude>
                <exclude>org/apache/drill/exec/rpc/control/**</exclude>
-               <exclude>org/apache/drill/exec/work/**</exclude>
```
Comment (adeneche) [inline exec/jdbc-all/pom.xml]: can you check with @dsbos or rahul if the JDBC test tool runs fine with this change ?

File: exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillConnectionImpl.java [hunk 8]
```diff
@@ -369,24 +370,33 @@ AvaticaFactory getFactory() {
     return factory;
   }
 
+  private static void closeOrWarn(final AutoCloseable autoCloseable, final String message, final Logger logger) {
```
Comment (adeneche) [inline exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillConnectionImpl.java]: should we make this public and move this to `org.apache.drill.common.AutoCloseables` ?

File: common/src/main/java/org/apache/drill/common/DrillAutoCloseables.java (hunk 3 above)
Comment (adeneche) [inline common/src/main/java/org/apache/drill/common/DrillAutoCloseables.java]: sounds good to me

File: exec/jdbc-all/pom.xml (hunk 7 above)
Comment (cwestin) [inline exec/jdbc-all/pom.xml]: I've sent them email, and copied you, Hakim.

File: exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillConnectionImpl.java (hunk 8 above)
Comment (cwestin) [inline exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillConnectionImpl.java]: Sounds good, but let's leave that until DrillAutoCloseables and AutoCloseables are merged, along with the change you suggested earlier/above.

File: exec/jdbc-all/pom.xml (hunk 7 above)
Comment (cwestin) [inline exec/jdbc-all/pom.xml]: Daniel checked it out and said it was fine.

**Comment by adeneche** [comment]
+1, LGTM

**Pull request closed without merging**